lab_2/main.py

```diff
@@ -0,0 +1,132 @@
+import requests
+import csv
+from datetime import datetime, timedelta
+from dateutil.relativedelta import relativedelta
+from collections import Counter
+import copy
+
+
+class MovieData:
+    API_PATH = 'https://api.themoviedb.org/3'
+    GENRE_URL = f'{API_PATH}/genre/movie/list?language=en'
+    DATA_URL = API_PATH + '/discover/movie?include_adult=false&include_video=false&sort_by=popularity.desc&page={!!!!}'
+    HEADERS = {
+        'accept': 'application/json',
+        'Authorization': 'Bearer eyJhbGciOiJIUzI1NiJ9.eyJhdWQiOiIzMTI3NGFmYTRlNTUyMjRjYzRlN2Q0NmNlMTNkOTZjOSIsInN1YiI6IjVkNmZhMWZmNzdjMDFmMDAxMDU5NzQ4OSIsInNjb3BlcyI6WyJhcGlfcmVhZCJdLCJ2ZXJzaW9uIjoxfQ.lbpgyXlOXwrbY0mUmP-zQpNAMCw_h-oaudAJB6Cn5c8'
+    }
+    FIELDNAMES = ['title', 'popularity', 'score', 'release_date', 'last_day_in_cinema']
+
+    def __init__(self, numb_of_pages):
+        self.genres = None
+        self.genre_dict = None
+        self.genre_names = None
+        self.pages = numb_of_pages
+        self.data = []
+
+    def fetch_all_data_in_pages(self):
+        for i in range(self.pages):
+            response = requests.get(self.DATA_URL, params={'page': i + 1}, headers=self.HEADERS)
+            self.data.extend(response.json()['results'])
+
+    def give_data(self):
+        return self.data
+
+    def give_data_with_indexes(self):
+        return self.data[3:19:4]
+
+    def find_popular_title(self):
+        return max(self.data, key=lambda x: x['popularity'])['original_title']
+
+    def find_film_with_keyword(self, keyword):
+        return [j['original_title'] for j in self.data if keyword in j['overview']]
+
+    def get_genres_data(self):
+        response = requests.get(self.GENRE_URL, headers=self.HEADERS)
+        self.genres = response.json()['genres']
+        self.genre_dict = {i['id']: i['name'] for i in self.genres}
+
+    def get_unique_collections(self):
+        all_genres = set()
+        genre_names = [self.genre_dict[j] for i in self.data for j in i['genre_ids'] if j in self.genre_dict]
+        all_genres.update(genre_names)
+        self.genre_names = {genre['id']: genre['name'] for genre in self.genres}
+        return all_genres
+
+    def delete_films_with_genre(self, genre_name_for_delete):
+        genre_id_for_delete = next(i['id'] for i in self.genres if i['name'].lower() == genre_name_for_delete.lower())
+        self.data = [i for i in self.data if genre_id_for_delete not in i['genre_ids']]
+        return self.data
+
+    def find_most_popular_genre(self):
+        all_genres = [genre_id for i in self.data for genre_id in i['genre_ids']]
+        popular_genres = [{'genre_name': self.genre_names[i], 'count': count}
+                          for i, count in Counter(all_genres).most_common()]
+        return popular_genres
+
+    def get_grouped_film_collection(self):
+        film_collections = [[film_1, film_2] for i, film_1 in enumerate(self.data) for film_2 in self.data[i + 1:]
```

Comment:
(the pairs should not allow inserts)
now your pairs is lists, and it allows inserts